From 538b008258573ff75618901b3c1696931aa05540 Mon Sep 17 00:00:00 2001
From: yangjunyi <yangjunyi5@huawei.com>
Date: Fri, 23 Dec 2022 10:20:13 +0800
Subject: [PATCH 0884/1398] [Huawei] drivers: hooks: add i40e driver vendor
 hooks

Offering: HULK
starry inclusion
category: bugfix
bugzilla: 187991

----------------------------------

Allow vendor modules to attach i40e driver hooks.
This patch introduce i40e hook example.

Usage:

  static void vendor_i40e_foo(void *data, unsigned int *size)
  {
          pr_info("%s\n", __func__);
  }

  static int __init vendor_init(void)
  {
	int ret = 0;
	ret += register_trace_vendor_i40e_example(&vendor_i40e_foo, NULL);

	return ret;
  }

  static void __exit vendor_exit(void)
  {
	unregister_trace_vendor_i40e_example(&vendor_i40e_foo, NULL);
  }

  module_init(vendor_init);
  module_exit(vendor_exit);

Signed-off-by: yangjunyi <yangjunyi5@huawei.com>
---
 drivers/hooks/Kconfig                       |  9 ++++
 drivers/hooks/vendor_hooks.c                |  6 +++
 drivers/net/ethernet/intel/i40e/i40e_main.c | 11 ++++-
 include/trace/hooks/i40e.h                  | 48 +++++++++++++++++++++
 4 files changed, 72 insertions(+), 2 deletions(-)
 create mode 100644 include/trace/hooks/i40e.h

diff --git a/drivers/hooks/Kconfig b/drivers/hooks/Kconfig
index 338bff92ef87..2ec1ab9af444 100644
--- a/drivers/hooks/Kconfig
+++ b/drivers/hooks/Kconfig
@@ -49,4 +49,13 @@ config VENDOR_IGB_HOOKS
               Allow vendor modules to attach igb driver hooks defined via
               DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
 
+config VENDOR_I40E_HOOKS
+        bool "I40e driver Vendor Hooks"
+        depends on VENDOR_HOOKS && I40E
+        default n
+        help
+            Enable i40e driver vendor hooks
+            Allow vendor modules to attach i40e driver hooks defined via
+            DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
+
 endmenu
diff --git a/drivers/hooks/vendor_hooks.c b/drivers/hooks/vendor_hooks.c
index 0b6e7c9aab49..39c6a3dbad40 100644
--- a/drivers/hooks/vendor_hooks.c
+++ b/drivers/hooks/vendor_hooks.c
@@ -12,6 +12,7 @@
 #include <trace/hooks/libata.h>
 #include <trace/hooks/spinor.h>
 #include <trace/hooks/igb.h>
+#include <trace/hooks/i40e.h>
 
 /*
  * Export tracepoints that act as a bare tracehook (ie: have no trace event
@@ -45,3 +46,8 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_ioctl);
 EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_set_spd_dplx);
 EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_phy_force_speed_duplex_setup);
 #endif
+
+#ifdef CONFIG_VENDOR_I40E_HOOKS
+EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_i40e_rebuild);
+EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_i40e_netdev_stats);
+#endif
diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index ea6a984c6d12..9ff453713dbe 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -6,6 +6,7 @@
 #include <linux/pci.h>
 #include <linux/bpf.h>
 #include <generated/utsrelease.h>
+#include <trace/hooks/i40e.h>
 
 /* Local includes */
 #include "i40e.h"
@@ -442,8 +443,13 @@ static void i40e_get_netdev_stats_struct(struct net_device *netdev,
 	struct i40e_ring *ring;
 	int i;
 
-	if (test_bit(__I40E_VSI_DOWN, vsi->state))
-		return;
+	if (test_bit(__I40E_VSI_DOWN, vsi->state)) {
+		unsigned int flag = 0;
+
+		trace_vendor_i40e_netdev_stats(&flag);
+		if (flag == 0)
+			return;
+	}
 
 	if (!vsi->tx_rings)
 		return;
@@ -9439,6 +9445,7 @@ static void i40e_reset_subtask(struct i40e_pf *pf)
 		i40e_prep_for_reset(pf, false);
 		i40e_reset(pf);
 		i40e_rebuild(pf, false, false);
+		trace_vendor_i40e_rebuild(pf);
 	}
 
 	/* If we're already down or resetting, just bail */
diff --git a/include/trace/hooks/i40e.h b/include/trace/hooks/i40e.h
new file mode 100644
index 000000000000..19c462a71b95
--- /dev/null
+++ b/include/trace/hooks/i40e.h
@@ -0,0 +1,48 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ * Description: I40e driver Vendor Hooks
+ */
+
+#ifdef CONFIG_VENDOR_I40E_HOOKS
+
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM i40e
+
+#define TRACE_INCLUDE_PATH trace/hooks
+#if !defined(_TRACE_HOOK_I40E_HOOKS_H) || defined(TRACE_HEADER_MULTI_READ)
+#define _TRACE_HOOK_I40E_HOOKS_H
+#include <linux/tracepoint.h>
+#include <trace/hooks/vendor_hooks.h>
+
+struct i40e_vsi;
+struct i40e_pf;
+struct i40e_hw;
+
+DECLARE_HOOK(vendor_i40e_rebuild,
+	TP_PROTO(struct i40e_pf *pf),
+	TP_ARGS(pf));
+
+DECLARE_HOOK(vendor_i40e_netdev_stats,
+	TP_PROTO(unsigned int *flag),
+	TP_ARGS(flag));
+
+#endif
+/* This part must be outside protection */
+#include <trace/define_trace.h>
+
+#else
+#include <linux/tracepoint.h>
+struct i40e_vsi;
+struct i40e_pf;
+struct i40e_hw;
+
+DECLARE_EVENT_NOP(vendor_i40e_rebuild,
+	TP_PROTO(struct i40e_pf *pf),
+	TP_ARGS(pf));
+
+DECLARE_EVENT_NOP(vendor_i40e_netdev_stats,
+	TP_PROTO(unsigned int *flag),
+	TP_ARGS(flag));
+
+#endif /* CONFIG_VENDOR_I40E_HOOKS */
-- 
2.17.1

